patchwork-fml/src/main/java/net/minecraftforge/forgespi/language/ModFileScanData.java

```diff
@@ -0,0 +1,250 @@
+
+			return annotationData;
+		}
+
+		private void initAnnotationData() {
+			annotationData = new HashMap<>();
+
+			try {
+				// TODO: This *may* load classes in the wrong order, but it shouldn't be an issue
+				Class<?> clazzObj = Class.forName(clazz.getClassName());
+				Class annotationType = Class.forName(this.annotationType.getClassName());
+				Annotation annotationObject = getAnnotationObject(clazzObj, annotationType);
+
+				if (annotationObject == null) {
+					LOGGER.error(String.format("Cannot fetch annotation object %s %s %s %s",
+							annotationType, targetType, clazz, memberName
+					));
+					return;
+				}
+
+				Method[] argMethods = annotationObject.getClass().getDeclaredMethods();
+
+				for (Method argMethod : argMethods) {
+					if (isArgumentMethod(argMethod)) {
+						annotationData.put(
+								argMethod.getName(),
+								argMethod.invoke(annotationObject)
+						);
+					}
+				}
+			} catch (Throwable e) {
+				LOGGER.catching(e);
+			}
+		}
+
+		private static boolean isArgumentMethod(Method method) {
+			String name = method.getName();
+			if (name.equals("toString")) return false;
+			if (name.equals("hashCode")) return false;
+			if (name.equals("getClass")) return false;
+			if (name.equals("equals")) return false;
+			if (name.equals("annotationType")) return false;
+
+			return true;
+		}
+
+		private Annotation getAnnotationObject(Class<?> clazzObj, Class annotationType) throws Throwable {
+			switch (targetType) {
+			case TYPE:
+				return clazzObj.getAnnotation(annotationType);
+			case FIELD:
+				return clazzObj.getField(memberName)
+						.getAnnotation(annotationType);
+			case METHOD:
+				//TODO handle overloaded methods
+				String methodName = memberName.substring(0, memberName.indexOf('('));
+				return Arrays.stream(clazzObj.getDeclaredMethods())
+						.filter(method -> method.getName().equals(methodName))
+						.findFirst()
+						.orElseThrow(() -> new RuntimeException("Cannot get method " + memberName))
+						.getAnnotation(annotationType);
+			default:
+				return null;
+			}
+		}
+
+		@Override
+		public boolean equals(final Object obj) {
+			if (!(obj instanceof AnnotationData)) {
+				return false;
+			}
+
+			AnnotationData dat = (AnnotationData) obj;
+			return Objects.equals(annotationType, dat.annotationType)
+					&& Objects.equals(targetType, dat.targetType)
+					&& Objects.equals(clazz, dat.clazz)
+					&& Objects.equals(memberName, dat.memberName);
+		}
+
+		@Override
+		public int hashCode() {
+			return Objects.hash(annotationType, targetType, clazz, memberName);
+		}
+	}
+
+	//if it's a jij mod, return parent mod's container
+	private static ModContainer getModContainer(String modid) {
+		ModContainer modContainer = FabricLoader.INSTANCE.getModContainer(modid)
+				.orElseThrow(() -> new RuntimeException("Cannot get mod container " + modid));
+
+		CustomValue customValue = modContainer.getMetadata().getCustomValue("modmenu:parent");
+
+		if (customValue == null) {
```

Comment:
How does this prove that a mod is JIJ or not? Anybody can set a parent to any mod, and a JIJ mod does not have to have a parent.

Reply:
Normal Fabric mod does not need annotation scanning, and patched mods have this added by the patcher.

Reply:
I'm not sure they do, but this would also match non-patchwork mods. It would be best to make a PR to patcher that adds a "patchwork:parent = foo" to the fabric mod json.

Reply:
How does Forge deal with this? Surely it doesn't try to assign annotation data to a single mod ID because it's a property of the mod file. Maybe we need a patchwork:annotation_holder attribute to hold the target mod ID.